### PR TITLE
[env/burn-staging] BM UI top bar volume control

### PR DESCRIPTION
**Pull request by yurii-lubynets**
Partially implements 
- https://github.com/sparkletown/internal-sparkle-issues/issues/1072

MVP version, a single mute button for ambient noises to toggle Animatemap sounds

sparkle-mute-button-01

### CODE REVIEW

File: src/components/atoms/VolumeControl/VolumeControl.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import React, { useCallback } from "react";
```
Comment (yurii-lubynets) [inline src/components/atoms/VolumeControl/VolumeControl.tsx:1]: initial volume control component. Needs to be styles per design https://www.figma.com/file/92gx2sTMvCc2W0NhgAG7pd/SparkleVerse-UI?node-id=1023%3A17756

File: src/components/molecules/NavBar/NavBar.tsx [hunk 2]
```diff
@@ -326,6 +347,47 @@ export const NavBar: React.FC<NavBarPropsType> = ({ hasBackButton = true }) => {
                     </S.RadioWrapper>
                   </S.RadioTrigger>
                 )}
+
+                <OverlayTrigger
+                  trigger="click"
+                  placement="bottom-end"
+                  overlay={
```
Comment (yurii-lubynets) [inline src/components/molecules/NavBar/NavBar.tsx:354]: update overlay prop as a separate component `<VolumePopover/>`

**Comment by yurii-lubynets** [comment]
Well, the main issue is to manage sounds from notifications and interactions. 
I've made:
P1 - Mute button for the background noise
P2 - Volume slider for the background noise

Styles need to be polished and created logic for manage volume

File: src/components/atoms/VolumeControl/VolumeControl.scss [hunk 3]
```diff
@@ -0,0 +1,45 @@
+@import "scss/constants";
```
Comment (goran-peoski-work) [inline src/components/atoms/VolumeControl/VolumeControl.scss:1]: extracted `VolumeControl` component out of `NavBar.tsx` was still using styles from `NavBar.scss`

File: src/components/atoms/VolumeControl/VolumeControl.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  className?: string;
+  label?: string;
+  name: string;
+  muted?: boolean;
+  onMute?: () => void;
+  onVolume?: (volume: number) => void;
+  volume?: number;
+  withMute?: boolean;
+  withSlider?: boolean;
+}
```
Comment (goran-peoski-work) [inline src/components/atoms/VolumeControl/VolumeControl.tsx:25]: A bit more props than currently used but left as stubs for future use cases.

File: src/components/molecules/NavBar/NavBar.scss [hunk 5]
```diff
@@ -1,6 +1,8 @@
 @import "scss/constants.scss";
 
-$border-radius: 28px;
+$menu-item-size: 32px;
+$volume-popover-width-with-sliders: 30em;
+$volume-popover-width-without-sliders: 15em;
```
Comment (goran-peoski-work) [inline src/components/molecules/NavBar/NavBar.scss:5]: Currently no sliders are being used, so when they land, one of these constants should be deleted and the other one used.

File: src/components/molecules/NavBar/NavBar.tsx [hunk 6]
```diff
@@ -140,7 +159,37 @@ export const NavBar: React.FC<NavBarPropsType> = ({ hasBackButton = true }) => {
...
+  const {
+    volume: notificationsVolume,
+    // volumeCallback: setNotificationsVolume,
+  } = useVolumeControl();
+
+  // Interactions volume control
+  const {
+    volume: interactionsVolume,
+    // volumeCallback: setInteractionsVolume,
+  } = useVolumeControl();
```
Comment (goran-peoski-work) [inline src/components/molecules/NavBar/NavBar.tsx:192]: Unused currently, left as stubs for future use.

**Review by goran-peoski-work — COMMENTED** [review]
Initial PR notes

File: src/settings.ts [hunk 7]
```diff
@@ -765,3 +765,14 @@ export const EVENT_STARTING_SOON_TIMEFRAME = 120; // in minutes
 export const EVENTS_PREVIEW_LIST_LENGTH = 5;
 
 export const BM_PARENT_ID = "playa";
+
+// NOTE: volume numbers are expressed in %, thus between 0 and 100
+export const DEFAULT_AUDIO_VOLUME: number = 10;
+export const DEFAULT_AMBIENT_VOLUME: number = 10;
+export const DEFAULT_INTERACTIONS_VOLUME: number = 50;
+export const DEFAULT_NOTIFICATIONS_VOLUME: number = 100;
```
Comment (sunny-viktoryia) [inline src/settings.ts:773]: looks like these constants are not used anywhere. do you think we should still leave there? the developer who is going to implement volume sliders might notice them and will create new ones.

File: src/hooks/useVolumeControl.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  const volumeCallback = useCallback(
+    (value: number) => {
+      setVolume(value);
+    },
+    [setVolume]
+  );
+
+  return {
+    volume,
+    volumeCallback,
```
Comment (sunny-viktoryia) [inline src/hooks/useVolumeControl.ts:19]: i guess we can add `mute` actions to the hook as well to have everything in one place.

File: src/hooks/useAudioVolume.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+export const useAudioVolume: (options: {
+  audioElement?: HTMLAudioElement;
+  initialVolume: number;
+  isAudioPlaying: boolean;
+  storageKey: string;
+}) => {
+  volume: number;
+  setVolume: (volume: number) => void;
+} = ({ audioElement, initialVolume, isAudioPlaying, storageKey }) => {
+  const [volume, setVolume] = useState<number>(initialVolume);
```
Comment (sunny-viktoryia) [inline src/hooks/useAudioVolume.ts:15]: should we use `useVolumeControl` here?

File: src/components/organisms/RadioModal/RadioModal.tsx [hunk 10]
```diff
@@ -34,15 +36,7 @@ export const RadioModal: React.FunctionComponent<RadioModalPropsType> = ({
...
-      <input
-        type="range"
-        id="vol"
-        name="vol"
-        min="0"
-        max="100"
-        onChange={(ev) => setVolume(Number(ev.target.value))}
-        value={volume}
-      />
+      <VolumeControl volume={volume} onVolume={setVolume} name="vol" />
```
Comment (sunny-viktoryia) [inline src/components/organisms/RadioModal/RadioModal.tsx:39]: is that correct that VolumeControl does not show anything 
image

previously it was this way:
image

File: src/components/atoms/VolumeControl/VolumeControl.tsx [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+      return onVolume?.(volume);
+    },
+    [onVolume]
+  );
+
+  const parentClasses = classNames({
+    VolumeControl: true,
+    VolumeControl__container: true,
+    [className]: className,
+  });
```
Comment (sunny-viktoryia) [inline src/components/atoms/VolumeControl/VolumeControl.tsx:50]: can we do something like this:
```suggestion
  const parentClasses = classNames("VolumeControl", { className });
```

why do we need 2 class names for the same div?

File: src/components/atoms/VolumeControl/VolumeControl.tsx [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  onMute?: () => void;
+  onVolume?: (volume: number) => void;
+  volume?: number;
+  withMute?: boolean;
+  withSlider?: boolean;
+}
+
+export const VolumeControl: React.FC<VolumeControlProps> = ({
+  className = "",
+  label = "",
```
Comment (sunny-viktoryia) [inline src/components/atoms/VolumeControl/VolumeControl.tsx:29]: do we have to set them? 
```suggestion
  className,
  label,
```

**Review by sunny-viktoryia — COMMENTED** [review]
partial review, as I'm not sure what should work and what shouldn't. feels like there is a lot of extra code.

Mute / unmute of "Ambient Noise" works!! 🎉 
couldn't see any volume sliders (for radio it disappeared as well) - was it done on purpose? 

@goran-peoski-work are you going to take the next task with those volume controls?

File: src/settings.ts [hunk 13]
```diff
@@ -765,3 +765,14 @@ export const EVENT_STARTING_SOON_TIMEFRAME = 120; // in minutes
...
+export const MIN_AUDIO_VOLUME: number = 0;
+export const MAX_AUDIO_VOLUME: number = 100;
+
+export const LS_KEY_IS_AMBIENT_AUDIO_VOCAL = "LS_KEY_IS_AMBIENT_AUDIO_VOCAL";
+export const LS_KEY_RADIO_VOLUME = "LD_KEY_RADIO_VOLUME";
```
[the remark this answers is not in the document]
Comment (goran-peoski-work) [reply src/settings.ts:778]: Typo, can be any string that's unique

File: src/store/reducers/AnimateMap.ts [hunk 14]
```diff
@@ -101,15 +103,16 @@ const initialAnimateMapState: AnimateMapState = {
   lastZoom: lastZoom ? parseFloat(lastZoom) : 1,
   //flags
   firstEntrance: window.sessionStorage.getItem("AnimateMapState.firstEntrance"),
-  environmentSound: true,
+  environmentSound:
+    window.localStorage.getItem(LS_KEY_IS_AMBIENT_AUDIO_VOCAL) !== "false",
```
[the remark this answers is not in the document]
Comment (goran-peoski-work) [reply src/store/reducers/AnimateMap.ts:107]: null by default can be true or false, comparing with `"false"` as a string ensures current behavior remains (default on)

File: src/hooks/useAudioVolume.ts [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+      console.error("useRadio", "Couldn't load initial volume", "error:", e);
+
+      Bugsnag.notify(e, (event) => {
+        event.addMetadata("context", {
+          location: "hooks/useRadio::useRadio",
+          storageKey: storageKey,
+          isInitialStateLoaded: isInitialStateLoadedRef.current,
+          volume,
+        });
+      });
```
[the remark this answers is not in the document]
Comment (goran-peoski-work) [reply src/hooks/useAudioVolume.ts:43]: Yes, any one liner code suggestions are welcome.

File: src/settings.ts (hunk 7 above)
Comment (goran-peoski-work) [reply src/settings.ts:773]: > looks like these constants are not used anywhere. do you think we should still leave there? the developer who is going to implement volume sliders might notice them and will create new ones.

I don't know what "might notice them and will create new ones" is, maybe a typo, but, as even from Yurii's code before, there are few things left for future expansion (alongside with sparse comments where they are used.

File: src/hooks/useVolumeControl.ts (hunk 8 above)
Comment (goran-peoski-work) [reply src/hooks/useVolumeControl.ts:19]: Good idea for future PR

File: src/hooks/useAudioVolume.ts (hunk 9 above)
Comment (goran-peoski-work) [reply src/hooks/useAudioVolume.ts:15]: I was asking myself the same questions, but was hesitant to remove stuff before I add the minimal required functionality for this PR

File: src/components/organisms/RadioModal/RadioModal.tsx (hunk 10 above)
Comment (goran-peoski-work) [reply src/components/organisms/RadioModal/RadioModal.tsx:39]: This one needs a prop flag. Good catch. I just don't have a reference venue to compare it before/after

File: src/components/atoms/VolumeControl/VolumeControl.tsx (hunk 11 above)
Comment (goran-peoski-work) [reply src/components/atoms/VolumeControl/VolumeControl.tsx:50]: I've shared this before: https://doesitsparkle.slack.com/archives/C01ER4M61TL/p1628775557030900 with the first reference being https://github.com/sparkletown/sparkle/pull/1705#discussion_r687619209

TL;DR:
One `className` prop, multiple classes per element, a single id per document element, like HTML+CSS is expected to have.

Longer form:
Two classes (not class names) are needed for the same reason Bootstrap uses both `btn` and `btn-primary` at the same element and for the same reason there is B in BEM :
 
a reference to the top level element that will make it easier to (re-)style elements inside components from the outside without having to research/update each time what their internal specifics are or if they have changed meanwhile.

File: src/components/atoms/VolumeControl/VolumeControl.tsx (hunk 12 above)
Comment (goran-peoski-work) [reply src/components/atoms/VolumeControl/VolumeControl.tsx:29]: Do we have to not set them?

**Comment by goran-peoski-work** [comment]
@sunny-viktoryia Two important things:

1. couldn't see any volume sliders (for radio it disappeared as well) - was it done on purpose?
Not on purpose. I think I just have to add `withSlider` button to have it reappear. I could use help by pointing me to a venue that has it enabled so I can check.

2. @goran-peoski-work are you going to take the next task with those volume controls?
Not sure, but whomever does continue ( on the next PR(s) ) will have my help and will probably have to wait on NB and check other stuff so the  volume controls are properly connected and functioning

All in all, this is a very early draft of a very larger issue that is considered P0 to be just a minimal viable (far from fully polished) addition for the sake of the mute button.

File: src/settings.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (goran-peoski-work) [reply src/settings.ts]: Added more explanation comments above @ eda678a6b658cc936def13246a4656c0b906b30d

File: src/store/reducers/AnimateMap.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (goran-peoski-work) [reply src/store/reducers/AnimateMap.ts:108]: Added a comment above the line noting why the check is done so. In short for null to default to true @ eda678a6b658cc936def13246a4656c0b906b30d

File: src/hooks/useAudioVolume.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (goran-peoski-work) [reply src/hooks/useAudioVolume.ts]: Fixed strings to use `useAudioVolume` as a more generic term than `useRadio` @ eda678a6b658cc936def13246a4656c0b906b30d

File: src/settings.ts (hunk 7 above)
Comment (goran-peoski-work) [reply src/settings.ts:774]: I will leave them as they are since there are few more places where stubs for volume controls are put for future addition/expansion

File: src/components/organisms/RadioModal/RadioModal.tsx (hunk 10 above)
Comment (goran-peoski-work) [reply src/components/organisms/RadioModal/RadioModal.tsx]: The Return of the Jed... Radio Slider @ eda678a6b658cc936def13246a4656c0b906b30d

File: src/components/atoms/VolumeControl/VolumeControl.tsx (hunk 11 above)
Comment (goran-peoski-work) [reply src/components/atoms/VolumeControl/VolumeControl.tsx]: Brand new className, now with minues one E class and plus 3 M ones @ eda678a6b658cc936def13246a4656c0b906b30d

File: src/components/atoms/VolumeControl/VolumeControl.tsx (hunk 12 above)
Comment (goran-peoski-work) [reply src/components/atoms/VolumeControl/VolumeControl.tsx:29]: I left them as they are. It's simple to see defaults at the top before digging into the meat of the component

**Review by mike-lvov — APPROVED** [review]
Pushing this baby in
